source/tutorials/file-sets.md

```diff
@@ -0,0 +1,694 @@
+```
+
+However, if you run `nix-build` again, the output path will be different!
+
+```shell-session
+$ nix-build
+trace: /home/user/fileset (all files in directory)
+this derivation will be built:
+  /nix/store/nlh7ismrf27xsnl3m20vfz6rvwlbbbca-fileset.drv
+...
+'hello.txt' -> '/nix/store/xknflcvjaa8dj6a6vkg629zmcrgz10rh-fileset/hello.txt'
+'world.txt' -> '/nix/store/xknflcvjaa8dj6a6vkg629zmcrgz10rh-fileset/world.txt'
+...
+/nix/store/xknflcvjaa8dj6a6vkg629zmcrgz10rh-fileset
+```
+
+The problem here is that `nix-build` by default creates a `result` symlink in the working directory, which points to the store path just produced:
+
+```
+$ ls -l result
+result -> /nix/store/xknflcvjaa8dj6a6vkg629zmcrgz10rh-fileset
+```
+
+Since `src` refers to the whole directory, and its contents change when `nix-build` succeeds, Nix will have to start over every time.
+
+:::{note}
+This will also happen without the file set library, e.g. when setting `src = ./.;` directly.
+:::
+
+The [`difference`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.difference) function subtracts one file set from another.
+The result is a new file set that contains all files from the first argument that aren't in the second argument.
+
+Use it to filter out `./result` by changing the `sourceFiles` definition:
+
+```{code-block} diff
+:caption: build.nix
+ { stdenv, lib }:
+ let
+   fs = lib.fileset;
+-  sourceFiles = ./.;
++  sourceFiles = fs.difference ./. ./result;
+ in
+```
+
+Building this, the file set library will specify which files are taken from the directory:
+
+```shell-session
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+this derivation will be built:
+  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
+...
+'hello.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/hello.txt'
+'world.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/world.txt'
+...
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+An attempt to repeat the build will re-use the existing store path:
+
+```
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+## Missing files
+
+Removing the `./result` symlink creates a new problem, though:
+
+```shell-session
+$ rm result
+$ nix-build
+error: lib.fileset.difference: Second argument (negative set)
+  (/home/user/select/result) is a path that does not exist.
+  To create a file set from a path that may not exist, use `lib.fileset.maybeMissing`.
+```
+
+Follow the instructions in the error message, and use `maybeMissing` to create a file set from a path that may not exist (in which case the file set will be empty):
```

Comment:
Note that this is not yet merged: https://github.com/NixOS/nixpkgs/pull/265964

So to make this work for now, you'll have to use that branch instead:
```
nix-shell -p niv --run "niv update nixpkgs --owner tweag --branch fileset.optional"
```

We shouldn't merge this PR until at least that PR is merged.

Reply:
Unblocked